src/constructors/tessellation.c

```diff
@@ -0,0 +1,305 @@
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301 USA
+*/
+
+#include "igraph_constructors.h"
+
+/**
+ * \ingroup generators
+ * \function igraph_triangle_lattice
+ * \brief A triangle lattice with the given shape.
+ *
+ * Creates a triangle lattice whose vertices have the form (i, j) for non-negative integers i and j
+ * and (i, j) is generally connected with (i + 1, j), (i, j + 1), and (i - 1, j + 1).
+ *
+ * </para><para>
+ * The vertices of the resulting graph are ordered lexicographically with the 2nd coordinate being
+ *  more significant, e.g., (i, j) < (i + 1, j) and (i + 1, j) < (i, j + 1)
+ *
+ * \param graph An uninitialized graph object.
+ * \param directed Boolean, whether to create a directed graph.
+ *        If the \c mutual argument is  not set to true,
+ *        edges will be directed from lower-index vertices towards
+ *        higher-index ones.
+ * \param mutual Boolean, if the graph is directed this gives whether
+ *        to create all connections as mutual.
+ * \param row_lengths_vector Integer vector, defines the number of vertices with
+ *        the second coordinate equal to the index. The length of this vector must match
+ *        the length of \p row_start_vector. All coordinates must be non-negative.
+ * \param row_start_vector Integer vector, defines the leftmost coordinate of
+ *        the vertex with the second coordinate equal to the index.
+ *
+ * \return Error code:
+ *         \c IGRAPH_EINVAL: invalid (negative) length of row_lengths_vector does not match the length of the
+            row_start_vector.
+ *
+ * Time complexity:  O(|V|), where |V| is the number of vertices in the generated graph.
+ */
+static igraph_error_t triangle_lattice(igraph_t *graph, igraph_bool_t directed, igraph_bool_t mutual,
+                                       igraph_vector_int_t row_lengths_vector, igraph_vector_int_t row_start_vector)
+{
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+    igraph_integer_t row_count = igraph_vector_int_size(&row_lengths_vector);
+    igraph_integer_t no_of_nodes;
+    igraph_vector_int_t row_lengths_prefix_sum_vector;
+    igraph_integer_t i, j;
+
+    if (igraph_vector_int_size(&row_lengths_vector) != igraph_vector_int_size(&row_start_vector))
+    {
+        IGRAPH_ERRORF(
+            "Length of row_lengths_vector vector must match the length of the "
+            "row_start_vector (%" IGRAPH_PRId ").",
+            IGRAPH_EINVAL, igraph_vector_int_size(&row_start_vector));
+    }
+
+    for (i = 0; i < row_count; i++)
+    {
+        if (VECTOR(row_lengths_vector)[i] < 0)
+        {
+            IGRAPH_ERRORF(
+                "row_lengths_vector vector must have non-negative coordinates "
+                "(%" IGRAPH_PRId ").",
+                IGRAPH_EINVAL, VECTOR(row_lengths_vector)[i]);
+        }
+    }
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 0);
```

Comment:
I copied this from https://github.com/igraph/igraph/blob/2c0cebbaa6e8179f1c2ed79f9d189eb22dc08f38/src/constructors/regular.c#L398 
Perhaps, this has something to do with failed tests.